### ISSUE TITLE
Error in cv_booster$record_evals[["valid"]][[first_metric]]

### ISSUE THREAD

**Issue by kailex** [microsoft/LightGBM#3112]
## How you are using LightGBM?
I'm trying to implement custom objective and evaluation functions. Lightgbm works fine either with custom objective or evaluation function, but not both.

LightGBM component: 
* R package

## Environment info

* Operating System: Debian 10.4
* CPU/GPU model: Intel Core i5
* GCC: 8.3.0
* CMake version: 3.13.4
* R version: 4.0.0 
* LightGBM version or commit hash: 2.3.2

## Error message and / or logs
```
[1]:	val's error:0.0335196 
 Error in booster$record_evals[[first_valid_name]][[first_metric]] : 
  attempt to select less than one element in get1index 
```

## Reproducible example(s)

```R
library(lightgbm)

data(agaricus.train, package = "lightgbm")
data(agaricus.test, package = "lightgbm")
dtrain <- lgb.Dataset(agaricus.train$data, label = agaricus.train$label)
dtest <- lgb.Dataset.create.valid(dtrain, data = agaricus.test$data, label = agaricus.test$label)

param <- list(
  num_leaves = 4L
  , learning_rate = 1.0
  , objective = "binary"
)

logregobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1.0 / (1.0 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1.0 - preds)
  return(list(grad = grad, hess = hess))
}

evalerror <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1.0 / (1.0 + exp(-preds))
  err <- as.numeric(sum(labels != (preds > 0.5))) / length(labels)
  return(list(name = "error", value = err, higher_better = FALSE))
}

lgb.train(
  params = param
  , data = dtrain,
  , nrounds = 100,
  , valids = list(val = dtest)
  , early_stopping_rounds = 5
  , obj = logregobj
  , eval = evalerror
  , eval_freq = 10
)
```

## Steps to reproduce
Run the code above.

**Comment by jameslamb** [comment]
@kailex thanks for using `LightGBM` and for the reproducible example! I will take a look in the next day or two and get back to you

**Comment by jameslamb** [comment]
@kailex I think I've found the issue, and proposed a fix in #3117 . If you want to try it, you can install the R package from the branch in that pull request:

```shell
git clone git@github.com:jameslamb/LightGBM.git -b fix/custom-eva
cd LightGBM
Rscript build_r.R
```

**Comment by kailex** [comment]
@jameslamb I tried it and it seems that the issue is fixed.
